collector/collector.go

```diff
@@ -50,28 +54,100 @@ type ReportTemplate struct {
 	TestVersion string `json:"test_version"`
 }
 
+// PrivacySettings contains privacy settings for submitting measurements.
+type PrivacySettings struct {
+	// IncludeASN indicates whether to include the ASN
+	IncludeASN bool
+
+	// IncludeCountry indicates whether to include the country
+	IncludeCountry bool
+
+	// IncludeIP indicates whether to include the IP
+	IncludeIP bool
+}
+
+func (ps PrivacySettings) apply(m *model.Measurement) (err error) {
+	if ps.IncludeASN == false {
+		m.ProbeASN = model.DefaultProbeASNString
+	}
+	if ps.IncludeCountry == false {
+		m.ProbeCC = model.DefaultProbeCC
+	}
+	if ps.IncludeIP == false {
+		err = ps.maybeRewriteTestKeys(m) // MUST be before the following statement
+		m.ProbeIP = model.DefaultProbeIP
+	}
+	return
+}
+
+func (ps PrivacySettings) maybeRewriteTestKeys(m *model.Measurement) error {
+	if net.ParseIP(m.ProbeIP) == nil {
+		return errors.New("Invalid m.ProbeIP string")
+	}
+	data, err := json.Marshal(m.TestKeys)
+	if err != nil {
+		return err
+	}
+	// The check using Count is to save an unnecessary copy performed by
+	// ReplaceAll when there are no matches into the body. This is what
+	// we would like the common case to be, meaning that the code has done
+	// its job correctly and has not leaked the IP.
+	pip := []byte(m.ProbeIP)
+	if bytes.Count(data, pip) <= 0 {
+		return nil
+	}
+	// Killing the previous TestKeys in my intention reduces the memory
+	// pressure possibly caused by this operation. We have, in fact, two
+	// copies of the `data` slice plus the test keys. TBH, I'm not sure
+	// however that losing all references to the test keys here is enough
+	// that they are immediately cleared, or whether that happens later.
+	m.TestKeys = nil
+	data = bytes.ReplaceAll(data, pip, []byte(`[REDACTED]`))
+	// We add an annotation such that hopefully later we can measure the
+	// number of cases where we failed to sanitize properly.
+	m.AddAnnotation("_probe_engine_sanitize_test_keys", "true")
+	// We assume that our string replacement here cannot make the resulting
+	// serialized JSON non parseable. Since we are very sure about this, let's
+	// be bold and make the code crash if that happens.
+	if err := json.Unmarshal(data, &m.TestKeys); err != nil {
+		panic(err)
+	}
+	return nil
+}
+
+type openResponse struct {
+	ID string `json:"report_id"`
+}
+
 // Report is an open report
 type Report struct {
 	// ID is the report ID
-	ID string `json:"report_id"`
+	ID string
+
+	// client is the client that was used.
+	client *Client
 
-	// Client is the client that was used.
-	Client *Client
+	// privacy contains privacy settings.
+	privacy PrivacySettings
 }
 
 // OpenReport opens a new report.
 func (c *Client) OpenReport(
-	ctx context.Context, rt ReportTemplate,
+	ctx context.Context, rt ReportTemplate, ps PrivacySettings,
 ) (*Report, error) {
-	var report Report
+	var or openResponse
```

Comment:
In the old code, the server could overwrite some fields. This is undesirable. Use a separate struct for getting the result of the JSON API call and then copy the values into the report.